server/src/main/java/org/elasticsearch/cluster/service/MasterService.java

```diff
@@ -985,11 +1002,18 @@ void onClusterStateUnchanged(ClusterState clusterState) {
             }
         }
 
-        void onPublishFailure(FailedToCommitClusterStateException e) {
+        void onPublishFailure(Exception e) {
+            assert e instanceof FailedToCommitClusterStateException || e instanceof NotMasterException;
```

Comment:
nit: not that this can fail today, but in general it's worth including the exception like this:

```suggestion
            assert e instanceof FailedToCommitClusterStateException || e instanceof NotMasterException : e;
```